From ccec1404e2aaaa6a0b085527ac1e75d492766b3b Mon Sep 17 00:00:00 2001
From: "Steve MacLean, Qualcomm Datacenter Technologies, Inc"
 <sdmaclea@qti.qualcomm.com>
Date: Wed, 3 May 2017 22:21:18 +0000
Subject: [PATCH] [Arm64/Unix] Enable
 FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP

---
 clrdefinitions.cmake      |  3 +++
 src/vm/arm64/asmhelpers.S | 25 +++++++++++++++++++++----
 src/vm/arm64/stubs.cpp    | 13 +++++++++++++
 3 files changed, 37 insertions(+), 4 deletions(-)

diff --git a/clrdefinitions.cmake b/clrdefinitions.cmake
index c2493f0..cd3707a 100644
--- a/clrdefinitions.cmake
+++ b/clrdefinitions.cmake
@@ -181,6 +181,9 @@ if(CLR_CMAKE_PLATFORM_ARCH_AMD64 AND NOT WIN32)
   add_definitions(-DFEATURE_MANUALLY_MANAGED_CARD_BUNDLES)
   add_definitions(-DFEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP)
 endif(CLR_CMAKE_PLATFORM_ARCH_AMD64 AND NOT WIN32)
+if(CLR_CMAKE_PLATFORM_ARCH_ARM64 AND NOT WIN32)
+  add_definitions(-DFEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP)
+endif(CLR_CMAKE_PLATFORM_ARCH_AMD64 AND NOT WIN32)
 if(WIN32)
     add_definitions(-DFEATURE_VERSIONING_LOG)
 endif(WIN32)
diff --git a/src/vm/arm64/asmhelpers.S b/src/vm/arm64/asmhelpers.S
index 15b8057..4121db3 100644
--- a/src/vm/arm64/asmhelpers.S
+++ b/src/vm/arm64/asmhelpers.S
@@ -310,17 +310,17 @@ LOCAL_LABEL(shadowupdateend):
     ldp  x12, x13, [sp],#16
 #endif
 
-    // Branch to Exit if the reference is not in the Gen0 heap
+    // Branch to DoneCardTable if the reference is not in the Gen0 heap
     //
     PREPARE_EXTERNAL_VAR g_ephemeral_low, x12
     ldr  x12,  [x12]
     cmp  x15,  x12
-    blt  LOCAL_LABEL(Exit)
+    blt  LOCAL_LABEL(DoneCardTable)
 
     PREPARE_EXTERNAL_VAR g_ephemeral_high, x12
     ldr  x12, [x12]
     cmp  x15,  x12
-    bgt  LOCAL_LABEL(Exit)
+    bgt  LOCAL_LABEL(DoneCardTable)
 
     // Check if we need to update the card table
     PREPARE_EXTERNAL_VAR g_card_table, x12
@@ -328,11 +328,28 @@ LOCAL_LABEL(shadowupdateend):
     add  x15,  x12, x14, lsr #11
     ldrb w12, [x15]
     cmp  x12, 0xFF
-    beq  LOCAL_LABEL(Exit)
+    beq  LOCAL_LABEL(DoneCardTable)
 
 LOCAL_LABEL(UpdateCardTable):
     mov  x12, 0xFF 
     strb w12, [x15]
+
+LOCAL_LABEL(DoneCardTable):
+
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+    // Update the write watch table if necessary
+    PREPARE_EXTERNAL_VAR g_sw_ww_enabled_for_gc_heap, x12
+    ldrb w12, [x12]
+    cbz  x12, LOCAL_LABEL(Exit)
+    PREPARE_EXTERNAL_VAR g_sw_ww_table, x12
+    lsr  x15, x14, #0xc  // SoftwareWriteWatch::AddressToTableByteIndexShift
+    add  x12, x12, x15
+    ldrb w15, [x12]
+    cbnz x15, LOCAL_LABEL(Exit)
+    mov  x15, 0xFF
+    strb w15, [x12]
+#endif
+
 LOCAL_LABEL(Exit):
     add  x14, x14, 8
     ret  lr  
diff --git a/src/vm/arm64/stubs.cpp b/src/vm/arm64/stubs.cpp
index 0c7eb4d..40d2749 100644
--- a/src/vm/arm64/stubs.cpp
+++ b/src/vm/arm64/stubs.cpp
@@ -1317,6 +1317,19 @@ void StompWriteBarrierResize(bool isRuntimeSuspended, bool bReqUpperBoundsCheck)
     return;
 }
 
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+void SwitchToWriteWatchBarrier(bool isRuntimeSuspended)
+{
+    return;
+}
+
+void SwitchToNonWriteWatchBarrier(bool isRuntimeSuspended)
+{
+    return;
+}
+#endif // FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+
+
 #ifdef DACCESS_COMPILE
 BOOL GetAnyThunkTarget (T_CONTEXT *pctx, TADDR *pTarget, TADDR *pTargetMethodDesc)
 {
-- 
2.7.4

